### PR TITLE
initial build-out

### PR DESCRIPTION
**Pull request by slpopejoy**
Code-complete implementations and tests of:
- `create-pair`
- `add-liquidity`
- `remove-liquidity`
- `swap-exact-in`
- `swap-exact-out`

### CODE REVIEW

File: pact/exchange.pact [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+  )
+
+  (defun add-liquidity:object
+    ( tokenA:module{fungible-v2}
+      tokenB:module{fungible-v2}
+      amountADesired:decimal
+      amountBDesired:decimal
+      amountAMin:decimal
+      amountBMin:decimal
+      account-id:string
```
Comment (mightybyte) [inline pact/exchange.pact:120]: This uses a single account name for both tokens.  Do we care that this prevents this function from working if the user doesn't have control of the same account on both tokens?

File: pact/exchange.pact [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+      allocs:[object{alloc}]
+    )
+    (require-capability (SWAPPING))
+    (let*
+      ( (head:object{alloc} (at 0 allocs))
+        (head-token:module{fungible-v2} (at 'token-out head))
+        (account (at 'account (at 'pair head)))
+        (out (at 'out head))
+      )
+      (head-token::transfer to account out)
```
Comment (mightybyte) [inline pact/exchange.pact:419]: Should this be `transfer-create` so the tx will fail if the guard doesn't match the guard that was passed in?

**Comment by fmelp** [comment]
Hey Stu, looked through the code thoroughly several times and cannot find anything specific to point out. The only thing that came to mind is a possible attack vector with maliciously implemented token fungible-v2 functions

**Comment by ggobugi27** [comment]
@slpopejoy  Could we add a function in `tokens.pact` to easily get a list of pairs that exist on chain?  Something like `(keys swap.tokens.ledger)` would be nice.

File: pact/exchange.pact (hunk 1 above)
Comment (slpopejoy) [reply pact/exchange.pact]: User has to have the same account on both tokens, similarly to Uniswap.

File: pact/exchange.pact (hunk 2 above)
Comment (slpopejoy) [reply pact/exchange.pact]: This is `transfer-create` now :+1:

**Comment by slpopejoy** [comment]
> @slpopejoy Could we add a function in `tokens.pact` to easily get a list of pairs that exist on chain? Something like `(keys swap.tokens.ledger)` would be nice.

@ggobugi27 added `get-tokens` 👍

**Review by ggobugi27 — CHANGES_REQUESTED** [review]
Looks really good!  The changes that I request is on `removeLiquidity`. and `swap-exact-out`.

File: pact/exchange.pact [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+    (enforce (and (> reserveA 0.0) (> reserveB 0.0)) "quote: insufficient liquidity")
+    (/ (* amountA reserveB) reserveA)
+  )
+
+
+  (defun remove-liquidity:object
+    ( tokenA:module{fungible-v2}
+      tokenB:module{fungible-v2}
+      liquidity:decimal
+      amountAMin:decimal
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply pact/exchange.pact:211]: It's an oversight missing from the port enforcing the minimums, I'll add

File: pact/exchange.pact [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+    (require-capability (UPDATING))
+    (with-capability (UPDATE pair-key reserve0 reserve1)
+      (update pairs pair-key
+        { 'leg0: { 'token: (at 'token (at 'leg0 p))
+                 , 'reserve: reserve0 }
+        , 'leg1: { 'token: (at 'token (at 'leg1 p))
+                 , 'reserve: reserve1 }}))
+  )
+
+  (defun add-liquidity:object
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply pact/exchange.pact:115]: I might leave this alone for now as the code is correct per the port from Solidity

File: pact/exchange.pact (hunk 3 above)
[the remark this answers is not in the document]
Comment (slpopejoy) [reply pact/exchange.pact:211]: added with tests

**Review by ggobugi27 — APPROVED** [review]

**Pull request merged**